[Bug Report] [TypeScript] [checkbox, checkbox-group] change事件响应方法类型约束是否不够严谨?

<!-- generated by https://issue.element-plus.org/ DO NOT REMOVE -->

Bug Type: **`TypeScript`**

## Environment

- Vue Version: `3.2.33`
- Element Plus Version: `2.2.4`
- Browser / OS: `Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.61 Safari/537.36`
- Build Tool: `Vite`

## Reproduction

### Related Component

- `el-checkbox`
- `el-checkbox-group`

### Reproduction Link

[Github Repo](https://github.com/wu67/yys-tool)

### Steps to reproduce

1. pull一下我的[小项目](https://github.com/wu67/yys-tool)
2. 把element-plus 更新至 2.2.0以上(不含2.2.0), 例如目前最新的2.2.4
3. npm run build

### What is Expected?

构建成功

### What is actually happening?

```bash
error TS2322: Type '(val: boolean) => void' is not assignable to type '(val: string | number | boolean) => any'.
  Types of parameters 'val' and 'val' are incompatible.
    Type 'string | number | boolean' is not assignable to type 'boolean'.

20           @change="handleCheckAllChange"
              ~~~~~~
...
```

### Additional comments


1. checkbox使用过程中, 绑定值必定是 boolean number string中的一种, 但是目前最新版本的类型断言强制约束了需要同时处理3中类型? 或者得写成 onChange(arg: any) => void
2. checkbox-group 的change事件响应方法貌似是沿用了checkbox的约束, 这个方法获取到的参数明明是个数组...

<!-- 如果你需要插入截图，请在这粘贴。 -->

--- a/packages/components/checkbox/src/checkbox-group.vue
+++ b/packages/components/checkbox/src/checkbox-group.vue
@@ -17,7 +17,7 @@ import { UPDATE_MODEL_EVENT } from '@element-plus/constants'
 import { debugWarn } from '@element-plus/utils'
 import { useNamespace, useSize } from '@element-plus/hooks'
 import {
-  checkboxEmits,
+  checkboxGroupEmits,
   useCheckboxGroup,
   useCheckboxGroupId,
   useCheckboxGroupProps,
@@ -28,7 +28,7 @@ defineOptions({
 })
 
 const props = defineProps(useCheckboxGroupProps)
-const emit = defineEmits(checkboxEmits)
+const emit = defineEmits(checkboxGroupEmits)
 
 const { elFormItem } = useCheckboxGroup()
 const { groupId, isLabeledByFormItem } = useCheckboxGroupId(props, {

--- a/packages/components/checkbox/src/checkbox.ts
+++ b/packages/components/checkbox/src/checkbox.ts
@@ -3,7 +3,13 @@ import { toTypeString } from '@vue/shared'
 import { UPDATE_MODEL_EVENT } from '@element-plus/constants'
 import { formContextKey, formItemContextKey } from '@element-plus/tokens'
 import { useFormItemInputId, useSize, useSizeProp } from '@element-plus/hooks'
-import { debugWarn, isBoolean, isNumber, isString } from '@element-plus/utils'
+import {
+  debugWarn,
+  isArray,
+  isBoolean,
+  isNumber,
+  isString,
+} from '@element-plus/utils'
 import type { ComponentInternalInstance, ExtractPropTypes, PropType } from 'vue'
 import type { FormContext, FormItemContext } from '@element-plus/tokens'
 import type { ICheckboxGroupInstance } from './checkbox.type'
@@ -303,11 +309,18 @@ const useEvent = (
   }
 }
 
+type CheckboxValueType = string | number | boolean
+
 export const checkboxEmits = {
-  [UPDATE_MODEL_EVENT]: (val: string | number | boolean) =>
+  [UPDATE_MODEL_EVENT]: (val: CheckboxValueType) =>
     isString(val) || isNumber(val) || isBoolean(val),
-  change: (val: string | number | boolean) =>
+  change: (val: CheckboxValueType) =>
     isString(val) || isNumber(val) || isBoolean(val),
+}
+
+export const checkboxGroupEmits = {
+  [UPDATE_MODEL_EVENT]: (val: CheckboxValueType[]) => isArray(val),
+  change: (val: CheckboxValueType[]) => isArray(val),
 }
 
 export const useCheckbox = (

